### ISSUE TITLE
lang: Anchor IDL should be uploaded to program binary instead of separate PDA

### ISSUE THREAD

**Issue by ngundotra** [coral-xyz/anchor#3225]
To date, only 25% of Anchor programs have uploaded their IDL 🤯  (1,324 programs out of 5,335 anchor programs currently on `mainnet`).

If anchor IDLs were uploaded with program binaries, then we would have 100% IDL adoption, leading to increased transparency on Solana.

Instead of requiring developers to write the compressed IDL bytes to the IDL address, `anchor build` should write the zipped bytes to an ELF section called `.anchor.idl`. Example: https://github.com/nifty-oss/asset/blob/main/include-idl/src/lib.rs

Developers should still be able to opt-out of this default behavior and deploy the IDLs the old way.

**Comment by acheroncrypto** [comment]
Overall, this would be great.

> If anchor IDLs were uploaded with program binaries, then we would have 100% IDL adoption, leading to increased transparency on Solana.

It would increase the percentage, but I don't think we'd ever see 100% IDL adoption, especially if we give people the chance to opt out of this behavior.

> Developers should still be able to opt-out of this default behavior and deploy the IDLs the old way.

Is this for backwards compatibility? We have to get rid of the IDL instructions when the IDL exists in the binary, otherwise the binary size will grow significantly.

**Comment by ngundotra** [comment]
> It would increase the percentage, but I don't think we'd ever see 100% IDL adoption, especially if we give people the chance to opt out of this behavior.

Yes this is true. I think aiming for 50% adoption or higher over next year would be attainable, with this change alone.

On backwards compatibility, I am agnostic. 

On removing IDL instructions from the binary, this is super exciting because then deploying IDLs with your binary is essentially "free". In other words - this shouldn't change deployment costs 😄 ! That should make adopting this easier🤞

**Comment by ngundotra** [comment]
Core does not recommend putting IDLs in program binary. They are trying to shave all bits of program binaries down as much as possible.

Accordingly, working on a different approach. Foundation is working on getting a program to upload program metadata on-chain like IDL, `solana-verify` args, and more. It will handle buffer history versioning too. It will be immutable ASAP.

**Comment by ngundotra** [comment]
Here is roughly the program spec we are thinking about: https://solanafoundation.notion.site/Program-Metadata-128d36dad52d801e974fcc3a30736506

**Comment by ngundotra** [comment]
Okay, after further discussion with Anza VM devs [here](https://discord.com/channels/428295358100013066/967028962746327060/1301659560716669041), it seems that this was a miscommunication. Core seems to be fine with having arbitrary metadata in the ELF section headers. 

Given how much friction is caused by uploading IDLs via `IdlInstruction`s, I think uploading `IDL`s in section headers is the better way to go.

@febo and I have published `solana-include-idl` and `solana-include-idl-cli`  https://github.com/nifty-oss/asset/blob/main/include-idl/README.md under solanafoundation are testing it out with OreV2 & some other programs. We will be working to support this new place for IDLs in the Explorer over the next 2-3 weeks.

An additional major benefit of uploading IDLs into binaries is that we can avoid the social consensus issue of getting program developers to use an IDL program. It's not impossible, but would suck to manage.

**Comment by acheroncrypto** [comment]
If the goal is adoption, why not include the IDL transactions to the `anchor deploy` and `solana program deploy` commands to make it opt-out rather than opt-in?

Requiring to fetch the whole program data account just to read a small portion of it is likely to increase the load for sites like explorers quite a bit, especially when compared to just fetching the PDA account, which pretty much only contains the required data. Maybe we need a new RPC method to fetch IDLs from program data accounts?

I'm still fine with it though, just noting some downsides/concerns.

**Comment by ngundotra** [comment]
Adding IDL transactions to `anchor deploy` makes sense, I thought it used to be included FWIW 

Adding IDL transactions to `solana program deploy` is theoretically fine, but requires one of 2 difficult options:
1. Either every program must have `IdlInstruction`s (injected by Anchor, or manually added to native programs)
2. Or we have some universal IDL program & we have to work on migration and education

I don't think either of those options are particularly great solutions compared to shoving ELF metadata into program headers for now.

>Requiring to fetch the whole program data account just to read a small portion of it is likely to increase the load for sites like explorers quite a bit

Yup this is true. I'm working on more efficient parsers. Worst case, we host a public API to serve IDLs to the Explorer.

>Maybe we need a new RPC method to fetch IDLs from program data accounts?

Core said "no way Jose" to new RPC methods until RPC v2. Probably best solution to host an API.

Thanks for the feedback Acheron!

**Comment by Henry-E** [comment]
There's a new IDL program, can't remember the issue number but jacob mentioned moving to that as the ultimate solution.
